### PR TITLE
fix silent type promition for sparse COO tensors with `select`

### LINKED ISSUE #82150: Select coo tensor with zero NNZ does not preserve dtype (for integer types?)
Index select on sparse COO tensors with zero nnz values and integral dtypes results in the reduced dimension result having dtype `int64` always. 

### 🐛 Describe the bug

```python
import torch
dtypes = [torch.bool, torch.int8, torch.int32, torch.bfloat16, torch.float32, torch.float64]
for dtype in dtypes:
    a = torch.ones((6, 1), dtype=dtype)
    a_s = a.to_sparse_coo()
    if a_s[0].dtype != dtype:
        print(f'2d->1d index select (with nnz != 0) result dtype incorrect, for dtype {dtype}, got {a_s[0].dtype}')
    if a_s[0,0].dtype != dtype:
        print(f'2d->scalar index select (with nnz != 0) result dtype incorrect, for dtype {dtype}, got {a_s[0,0].dtype}')

    b_s = (a * 0).to_sparse_coo()
    if b_s[0].dtype != dtype:
        print(f'2d->1d index select (with nnz == 0) result dtype incorrect, for dtype {dtype}, got {b_s[0].dtype}')
    if b_s[0,0].dtype != dtype:
        print(f'2d->scalar index select (with nnz == 0) result dtype incorrect, for dtype {dtype}, got {b_s[0,0].dtype}')

```
### Output
```
2d->1d index select (with nnz == 0) result dtype incorrect, for dtype torch.bool, got torch.int64
2d->scalar index select (with nnz == 0) result dtype incorrect, for dtype torch.bool, got torch.int64
2d->scalar index select (with nnz == 0) result dtype incorrect, for dtype torch.int8, got torch.int64
2d->scalar index select (with nnz == 0) result dtype incorrect, for dtype torch.int32, got torch.int64
```

### Versions

Repro generated with current master

cc @nikitaved @pearu @cpuhrsch @amjames @bhosmer @nairbv @mruberry

### PR DESCRIPTION
**Pull request by nikitaved**
Fixes https://github.com/pytorch/pytorch/issues/82150.

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -1329,7 +1329,8 @@ static Tensor select_sparse(const Tensor& self, int64_t dim, int64_t index) {
       if (new_values.size(0) == 1) {
         return new_values[0];
       } else {
-        return new_values.sum(0);
+        // sum promotes integral type to int64 when dtype is not specified.
```
Comment (amjames) [inline aten/src/ATen/native/TensorShape.cpp:1332]: Seems like the default should be to use `self.scalar_type` but good catch!

**Review by amjames — APPROVED** [review]
Thanks for the quick turnaround on this!

File: aten/src/ATen/native/TensorShape.cpp (hunk 1 above)
Comment (nikitaved) [reply aten/src/ATen/native/TensorShape.cpp:1332]: Yeah, it is a bit unclear, but intentional. The doc could have been on that one, for sure!

**Review by cpuhrsch — APPROVED** [review]

**Comment by cpuhrsch** [comment]
You can rebase and try to merge again, maybe that'll fix the trunk failures.

**Comment by nikitaved** [comment]
Let me give a shot now :)

@pytorchbot merge -g

**Pull request closed without merging**